src/builder.rs

```diff
@@ -1016,50 +1280,50 @@ impl<V> Bundle<InProgress<Proof, Unauthorized>, V> {
     }
 }
 
-impl<P: fmt::Debug, V> Bundle<InProgress<P, PartiallyAuthorized>, V> {
+impl<Proof: fmt::Debug, V, P: OrchardPrimitives>
+    Bundle<InProgress<Proof, PartiallyAuthorized>, V, P>
+{
     /// Signs this bundle with the given [`SpendAuthorizingKey`].
     ///
-    /// This will apply signatures for all notes controlled by this spending key.
+    /// This will apply versioned signatures for all notes controlled by this spending key.
     pub fn sign<R: RngCore + CryptoRng>(self, mut rng: R, ask: &SpendAuthorizingKey) -> Self {
         let expected_ak = ask.into();
         self.map_authorization(
             &mut rng,
             |rng, partial, maybe| match maybe {
                 MaybeSigned::SigningMetadata(parts) if parts.ak == expected_ak => {
-                    MaybeSigned::Signature(
+                    MaybeSigned::Signature(VerSpendAuthSig::new(
+                        P::default_sighash_version(),
                         ask.randomize(&parts.alpha).sign(rng, &partial.sigs.sighash),
-                    )
+                    ))
                 }
                 s => s,
             },
             |_, partial| partial,
         )
     }
-    /// Appends externally computed [`Signature`]s.
+    /// Appends externally computed versioned signatures.
     ///
-    /// Each signature will be applied to the one input for which it is valid. An error
-    /// will be returned if the signature is not valid for any inputs, or if it is valid
+    /// Each versioned signature will be applied to the one input for which it is valid. An error
+    /// will be returned if the versioned signature is not valid for any inputs, or if it is valid
     /// for more than one input.
     ///
-    /// [`Signature`]: redpallas::Signature
-    pub fn append_signatures(
-        self,
-        signatures: &[redpallas::Signature<SpendAuth>],
-    ) -> Result<Self, BuildError> {
+    /// [`Signature`]: VerSpendAuthSig
+    pub fn append_signatures(self, signatures: &[VerSpendAuthSig]) -> Result<Self, BuildError> {
         signatures.iter().try_fold(self, Self::append_signature)
     }
 
-    fn append_signature(
-        self,
-        signature: &redpallas::Signature<SpendAuth>,
-    ) -> Result<Self, BuildError> {
+    fn append_signature(self, signature: &VerSpendAuthSig) -> Result<Self, BuildError> {
         let mut signature_valid_for = 0usize;
         let bundle = self.map_authorization(
             &mut signature_valid_for,
             |valid_for, partial, maybe| match maybe {
                 MaybeSigned::SigningMetadata(parts) => {
                     let rk = parts.ak.randomize(&parts.alpha);
-                    if rk.verify(&partial.sigs.sighash[..], signature).is_ok() {
+                    if rk
+                        .verify(&partial.sigs.sighash[..], signature.sig())
```

Comment:
Bug: This needs to check that the `sighash_info` within `signature` is compatible with `partial.sigs.sighash`. Currently it will either fall through and error with `BuildError::InvalidExternalSignature` (if `signature` was created correctly with a non-v0 sighash), or silently append a `signature` using the v0 sighash but a non-v0 `sighash_info`.

In practice I think this means that we will eventually need a different API for providing signatures over non-v0 sighashes (and maybe this actually only lives in PCZT-land for simplicity), so here we should return an error if the `sighash_info` within `signature` does not select the v0 sighash.